esmvalcore/_recipe.py

```diff
@@ -352,71 +352,86 @@ def _get_default_settings(variable, config_user, derive=False):
 def _add_fxvar_keys(fx_var_dict, variable):
     """Add keys specific to fx variable to use get_input_filelist."""
     fx_variable = dict(variable)
+    fx_variable.update(fx_var_dict)
 
     # set variable names
     fx_variable['variable_group'] = fx_var_dict['short_name']
-    fx_variable['short_name'] = fx_var_dict['short_name']
 
-    # specificities of project
+    # add special ensemble for CMIP5 only
     if fx_variable['project'] == 'CMIP5':
-        fx_variable['mip'] = 'fx'
         fx_variable['ensemble'] = 'r0i0p0'
-    elif fx_variable['project'] == 'CMIP6':
-        fx_variable['grid'] = variable['grid']
-        if 'mip' in fx_var_dict:
-            fx_variable['mip'] = fx_var_dict['mip']
-    elif fx_variable['project'] in ['OBS', 'OBS6', 'obs4mips']:
-        fx_variable['mip'] = 'fx'
+
     # add missing cmor info
     _add_cmor_info(fx_variable, override=True)
+
     return fx_variable
 
 
-def _get_correct_fx_file(variable, fx_varname, config_user):
+def _get_correct_fx_file(variable, fx_variable, config_user):
     """Get fx files (searching all possible mips)."""
-    # TODO: allow user to specify certain mip if desired
+    # make it a dict
+    if not isinstance(fx_variable, dict):
+        fx_varname = fx_variable
+        fx_variable = {'short_name': fx_variable}
+    else:
+        fx_varname = fx_variable['short_name']
+
+    # assemble info from master variable
     var = dict(variable)
     var_project = variable['project']
+    get_project_config(var_project)
     cmor_table = CMOR_TABLES[var_project]
+    valid_fx_vars = []
 
     # Get all fx-related mips ('fx' always first, original mip last)
     fx_mips = ['fx']
     fx_mips.extend(
         [key for key in cmor_table.tables if 'fx' in key and key != 'fx'])
     fx_mips.append(variable['mip'])
 
+    # force only the mip declared by user
+    if 'mip' in fx_variable:
+        fx_mips = [fx_variable['mip']]
```

Comment:
@schlunma care to tell Klaus the logic behind? I can say that this is a safeguard to account for all possible permutations

Reply:
> 
> 
> I also wonder if we can take the non-fx tables into account here: If the variable does not live in, eg, `Ofx`, but `Omon` then that will also be the `mip` of `variable`.

I'm not 100% sure if I got you correctly. Are you proposing to just search in every table instead of only the one related to `variable`? Right now, this only searches in the non-fx table with the `mip` of `variable`.

Reply:
...and that's about as safe as one can go at the moment, w/o going through all the tables, as Manuel points out

Reply:
There are two things here: One ist the code suggestion and the other is my comment. Let's deal with the suggestion first. Do you agree that it is functionally equivalent?

Reply:
OK plugged the code change :+1:

Reply:
Right, on to the comment. Maybe I misunderstand the code, but here is how I read it:

If the user has explicitly provided a `mip`(table) for the "fx" variable, use that. Otherwise, try to find the requested "fx" variable in all tables of the project (CMIP5 or CMIP6) that contain the literal string "fx", starting with `fx` itself.
If neither happens, presumably there will be an error down the line.

Now my suggestion stems from the observation that the most reasonable (maybe the only) other candidate for the table is the table of the data variable itself. For example, `volcello` for `(Omon, thetao)` can either come from `Omon` or `Ofx`. `Odec` would be unreasonable.
Hence, why not start the list of guessing tables with the table of the data variable?

Does that make sense?

Reply:
Yes, I agree, haven't thought about this thoroughly.

So I guess we can start the `else` loop with

```python
fx_mips = [variable['mip'], 'fx']
```

and drop the last line in the loop.

Reply:
Ok, I'll add a new suggestion to that effect.